### ISSUE TITLE
Project structure

### ISSUE THREAD

**Issue by londumas** [igmhub/SaclayMocks#16]
As discussed before, there few things to do before this code should be linked to a paper:
 - move the code from `gitlab.in2p3.fr` to `github.com`.
 - Change the code name so it is all lowercase
 - remove all upper case in file names everywhere
 - clean the repository `bin`, move `bin/data` to `data` or something of the sort...
The `bin` repository is for executable. I can't believe all >20 files are executables.
if they are library like, they should go to py/LyaMocks/
 - fill the readme with more explanation and an example of a run

**Comment by londumas** [comment]
@TEtourneau, thanks for the commit with all the code. I added Codacy to the code and it gives the following report: https://app.codacy.com/app/londumas/SaclayMocks/issues/index. Please have a look. I fixed the trailing white-spaces and the Readme in PR https://github.com/igmhub/SaclayMocks/pull/23 and PR https://github.com/igmhub/SaclayMocks/pull/24.
For the rest, I would rather have you do it since you know the code, and I don't.

**Comment by londumas** [comment]
@TEtourneau, I see that you also have a lot of commented lines, like in https://github.com/igmhub/SaclayMocks/blob/a1927f1ad787e0229d56952980380f4e0f8aab6a/bin/dla_saclay.py#L71 or https://github.com/igmhub/SaclayMocks/blob/a1927f1ad787e0229d56952980380f4e0f8aab6a/bin/make_boxes.py#L79. This can lead to very long files and unstable behavior. Consider getting rid of them if possible.

**Comment by londumas** [comment]
@TEtourneau, sending the following does not work on the Utah cluster:
```
python ~/Programs/igmhub/SaclayMocks/bin/submit_mocks.py --mock-dir ./ --cori-nodes False --box-size 256
WARNING: You are using Burst Buffer mode. If you cancel the sbatch jobs
or if there is any job bug, 
please cancel the burst buffer persistent reservation by running the script
delete_reservation.sh
--mock-dir option should point to /global/cscratch1 when using the burst
buffer mode !
```

**Comment by londumas** [comment]
@TEtourneau, thank for the update, I got through the previous error, and have a new one.
It is linked to python2.7 vs. python3 division: `npixel/100`.
Which python do you use? I use python3.
```
python ~/Programs/igmhub/SaclayMocks/bin/submit_mocks.py --mock-dir ./ --cori-nodes False --box-size 256
Warning: the jobs will not be sent to cori nodes, they will be executed here.
Writting scripts for 1 realisations
Mock files will be written in ./
It will run:
- Pk
- Boxes
- Chunks: qso randoms make_spectra merge_spectra 
- Merge Chunks: merge_qso merge_randoms compute_dla merge_dla dla_randoms transmissions 
It will produce 6 chunks, with ids: ['1' '2' '3' '4' '5' '6']
and ra0: ['189.982649735' '183.642649735' '177.302649735' '170.962649735'
 '196.322649735' '202.662649735']
Generating script for Pk...

Generating scripts for realisation 0
Creating directory tree...
Traceback (most recent call last):
  File "HOME/Programs/igmhub/SaclayMocks/bin/submit_mocks.py", line 1053, in <module>
    main()
  File "HOME/Programs/igmhub/SaclayMocks/bin/submit_mocks.py", line 1049, in main
    make_realisation(imock, mock_args, run_args, sbatch_args)
  File "HOME/Programs/igmhub/SaclayMocks/bin/submit_mocks.py", line 541, in make_realisation
    do_dir_tree(out_dir, mock_args['nside'])
  File "HOME/Programs/igmhub/SaclayMocks/bin/submit_mocks.py", line 420, in do_dir_tree
    for i in range(npixel/100 + 1):
TypeError: 'float' object cannot be interpreted as an integer
```

**Comment by TEtourneau** [comment]
Sorry not to have answer quicker.
I pushed the code but their were some bugs remaining anyway.

I corrected all the bugs and push the new code in the master. I did a try :
`submit_mocks.py --mock-dir . --cori-nodes False --box-size 256`
and seems to work. 

And yes, the code is written in python 2. I updated the README to mention this.
@londumas Could you try on Utah cluster ?

**Comment by londumas** [comment]
@TEtourneau, running it now. Let's have it Python3 compatible, because first Python2 is near decommission, and also Python3 is DESI standard. You can use `2to3` to track where you should update your code.

**Comment by TEtourneau** [comment]
I already tried to do this, but it's really a mess. And there are libraries that are not support anymore in python3.
I'll postpone this until it is really necessary.

**Comment by londumas** [comment]
@TEtourneau, ok.
I send the following and got the following.
```
<me> python ~/Programs/igmhub/SaclayMocks/bin/submit_mocks.py --mock-dir ./ --cori-nodes False --box-size 256
Warning: the jobs will not be sent to cori nodes, they will be executed here.
Writting scripts for 1 realisations
Mock files will be written in ./
It will run:
- Pk
- Boxes
- Chunks: qso randoms make_spectra merge_spectra 
- Merge Chunks: merge_qso merge_randoms compute_dla merge_dla dla_randoms transmissions 
It will produce 6 chunks, with ids: ['1' '2' '3' '4' '5' '6']
and ra0: ['189.982649735' '183.642649735' '177.302649735' '170.962649735'
 '196.322649735' '202.662649735']
Generating script for Pk...

Generating scripts for realisation 0
Creating directory tree...
Directory tree created.
.//mock_0/output/runs/submit.sh generated. Run it to submit the sbatch jobs of this realisation to the cori queue

```

I tried to run `.//mock_0/output/runs/submit.sh` but nothing happened. What should I do?

**Comment by TEtourneau** [comment]
Normally, submit.sh send the jobs to the cori nodes. If you're using the option --cori-nodes False, it will just run the script in your terminal, one after another.
There are no prints for this mode (I should add some).
You can check the logs of each script in ./mock_0/output/logs/
(The pk logs are stored in ./mock_0/pk/logs/, this is run before producing the chunks)
(you can see the order of the script in the readme)

**Comment by londumas** [comment]
@TEtourneau, ok thanks. some of the errors in the log are the following, what should I do?
```
gedit mock_0/output/logs/*log &
*** Running draw_qso 8 times ***
Error in run_draw_qso-0
==> Error in 1 run_draw_qso ...   Abort!
qso-1
Error in draw_qso-2
Error in draw_qso-3
Error in draw_qso-4
Error in draw_qso-5
Error in draw_qso-6
Error in draw_qso-7
==> Error in 8 draw_qso ...   Abort!
```

**Comment by londumas** [comment]
@TEtourneau, the code looks good. It is nice to have it on igmhub.
Here are things that would make it easier to read and run:
 - apply Codacy recommendations
 - rename all executable with `saclaymocks_` prefix, the same way we do for picca
 - have comments with same indentation as the code: e.g. https://github.com/igmhub/SaclayMocks/blob/011fe8556abfd7a89e46fd720cc8ff0f1f4d774b/py/SaclayMocks/box.py#L122
 - remove useless comments, like tests, print for tests or depreciated code
 - comments explaining what a function does should be between `"""` and `"""`
 - add a simple travis test, producing a smaa box and testing if recovering the same as the one stored. Same as we do in picca
Thank you for everything.

**Comment by londumas** [comment]
@TEtourneau, I just added the travis tests structure.
We now need you to add a test in `py/SaclayMocks/test/test_cor.py`. I would do a production of a very small box, stored in  `py/SaclayMocks/test/data/` and each PR would see if the box produced via the branch is the same as the one in the PR. How does that sound? You can have a look at how it is done in picca if you want.

**Comment by londumas** [comment]
@TEtourneau, let's continue discussions on travis tests here, because my branches seem to be always moving :)
Since I can not run the python2.7 tests, I would need you to run python test_cor.py on Nersc, changing
```
self._test = True
```
to 
```
self._test = False
self._branchFiles = './data_test/'
```
where you ran mkdir data_test
then tell me where I can read the files.

This will allow me to get the different reference master files, to test the travis against.

**Comment by londumas** [comment]
I have the following error in python2.7 tests:
 - `dist` is not set in merge_qso.py
 - `seed` seems random in pyigm
 - some warning on the range of P(k)
 - some warning on `etc/wisdom.128-128-1536.64.npy ` not existing

```
Reading QSO fits files...
Reading done. 0.013041973114 s
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/bin/merge_qso.py", line 4, in <module>
    __import__('pkg_resources').run_script('SaclayMocks==0.1', 'merge_qso.py')
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/pkg_resources/__init__.py", line 744, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1506, in run_script
    exec(script_code, namespace, namespace)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/SaclayMocks-0.1-py2.7.egg/EGG-INFO/scripts/merge_qso.py", line 154, in <module>
    
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/SaclayMocks-0.1-py2.7.egg/EGG-INFO/scripts/merge_qso.py", line 121, in main
    
NameError: global name 'dist' is not defined
Command exited with non-zero status 1
2.51Real 1.65user 0.16system 72%CPU 104236 
```

```
Using P14 spline values to generate a default model
Loading: /home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/pyigm/data/fN/fN_spline_z24.fits.gz
Seed has not been specified. Seed is set to 1212305484
Files will be read from /tmp/tmpaHgi0Z//Products//mock_0/chunk_1/spectra_merged/
Output will be written in /tmp/tmpaHgi0Z//Products//mock_0/chunk_1/dla.fits
Will read 20  files
Read 0 of 20
Fits table written.
Draw 6667 DLAs
Took 2.72931003571 s
5.40Real 4.18user 0.22system 81%CPU 149360 
```

```
InterpolatePk : 5th slice over 16.
**** WARNING rmax= 576.2510991117462  while rmax >= 1000 advised ****
produced  /tmp/tmp64sp4i30//Products//pk/P128-128-1536_5_16.fits
Took 8.656602621078491s
22.87Real 2.61user 0.23system 12%CPU 263740 
```

```
Starting DrawGRF...
Specified seed is 42
volume =  264328200.0
/home/travis/build/igmhub/SaclayMocks/py/SaclayMocks/../..//etc/wisdom.128-128-1536.64.npy  wisdomfile not found !!! set save_wisdom = true
128 128 1536
96.0  Mbytes box drawn 1.533642292022705  s
float32
FFT 10.70359182357788  s
/!\ Boxk was null. Wisdom has been saved, trying again FFTW...
Starting again DrawGRF_boxk...
Loading wisdom /home/travis/build/igmhub/SaclayMocks/py/SaclayMocks/../..//etc/wisdom.128-128-1536.64.npy
96.0  Mbytes box drawn 1.5283408164978027  s
```

**Comment by TEtourneau** [comment]
@londumas I corrected the bug in merge_qso, sorry for that ...
I also corrected the seed the dla_saclay, I just forgot to add the argument --seed in submit_mocks.py
Concerning the Pk warning: JM said that he will have a look to this, but it should be OK for the moment.
For the warning in make_boxes: it's just that the wisdom file for the FFT is not existing. I only store in etc/ the on for nominal boxes (2560-2560-1536 with 64 threads). So it had to create it

**Comment by TEtourneau** [comment]
Maybe we could store the wisdom file corresponding to the configuration of the travis test ? It will avoid this warning (and save some time)

**Comment by londumas** [comment]
@TEtourneau, thanks for the work. However you broke master: `https://travis-ci.org/igmhub/SaclayMocks/builds/530412647?utm_source=github_status&utm_medium=notification`. Please do not merge a branch until the travis tests are clear. Here is the error in master:
```
usage: dla_saclay.py [-h] --input_path INPUT_PATH --output_file OUTPUT_FILE
                     [--input_pattern INPUT_PATTERN] --fname_sigma FNAME_SIGMA
                     [--nmin NMIN] [--nmax NMAX] [--dla_bias DLA_BIAS]
                     [--cell_size CELL_SIZE] [--seed SEED]
dla_saclay.py: error: argument --seed: expected one argument
-----------------------------------------------------------
-----------------------------------------------------------
WARNING: Not loading mcmc in pyigm.fN   
 Install pymc if you want it
-----------------------------------------------------------
Using P14 spline values to generate a default model
Loading: /home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/pyigm/data/fN/fN_spline_z24.fits.gz
Command exited with non-zero status 2
2.47Real 1.35user 0.12system 59%CPU 87724 
```

**Comment by londumas** [comment]
@jmarclegoff and @TEtourneau, thanks for all the work.
I see in the log on the travis test that the `seed=42` is the same across all the different codes and across all the different processes. Is it the case, or actually you use the seed as a master seed to get different seeds for each processes? and each piece of code?

**Comment by TEtourneau** [comment]
Currently, if you fix the seed with
`--seed 42`
every python codes that is using numpy.random will have its seed fixed to 42.
So, for instance in the travis test, each of the draw_qso.py codes will run with a seed fixed to 42.
I guess it is not optimal, but it should do the trick for now.

**Comment by londumas** [comment]
@TEtourneau, ok thanks for the answer. At some point, we will need to change that. but it is ok for now.

**Comment by TEtourneau** [comment]
The fixing of the seed has been improved: now the seed of the job i is fixed to the specified seed (42 for instance) + the index of the i.
